### DISCUSSION TITLE
Removing Protobuf 2.5

### DISCUSSION THREAD

**Discussion by jojochuang** [apache/ozone#7565] [Ideas]
Hi I spent some time trying to get rid of protobuf 2.5. Took several approaches:

(1) Hadoop recommends migrating to ProtobufRpcEngine2 where it uses shaded protobuf classes org.apache.hadoop.thirdparty.protobuf.*. Tried to this approach but it's a huge mess. Not only does the RPC messages need to change to implement the shaded protobuf classes, rocksdb codec and column family must change as well. Grpc needs special treatment to use the new RPC messages too. What's worse, it implies we will be in lockstep with the Hadoop thirdparty protobuf version. That just doesn't seem right because Ozone don't talk to Hadoop or HDFS.

(2) Bump protobuf 2.5 to 3.x in the pom.xml, use protoc 3 instead of protoc 2.5 to compile protobuf message definition files. I can confirm that Hadoop's ProtobufRpcEngine works with protobuf 3. It doesn't require ProtobufRpcEngine2 to use protobuf3.
Advantage: minimum change.
Drawback: every test passes except with Hadoop 2 (ozonesecure-mr). In an Hadoop 2 environment, the class path does not have protobuf 3 jars. In addition, it probably won't work with Spark, or any other downstream application clients. (PoC: https://github.com/jojochuang/ozone/commits/refs/heads/remove_protobuf2.5_take2/)

(3) I've been thinking we should create a shaded uber filesystem jar, that in addition to approach (2), include & relocate Hadoop's ProtobufRpcEngine, protobuf3, Grpc (possibly via ozone-filesystem-shaded), so that everything related to com.google.protobuf get relocated to org.apache.ozone.shaded.com.google.protobuf. That way, it *should* work.
This approach has an additional benefit that Spark will no longer require ozone-filesystem-hadoop3-client to use Ozone. The problem observed in [HDDS-6926](https://issues.apache.org/jira/browse/HDDS-6926) is due to class path collision. If the client has ProtobufRpcEngine class relocated, it won't collide with the one coming from hadoop-client-api.

**Comment by adoroszlai** [comment]
(3): Seems too good to be true.  The main problem is that we cannot simply include and relocate Hadoop's `ProtobufRpcEngine` in itself.  It depends on and is used by other classes, which also need to be added.  The set of classes is Hadoop-version specific, so a shaded jar created for Hadoop 3.4.1 may not work with Hadoop 3.3, let alone 2.10.  It may also be dependent on usage/configuration, e.g. secure env. may require more classes.  On the other hand, we cannot just include and relocate everything from Hadoop, because we need to implement `FileSystem` and other interfaces in their original, unshaded version.  I think this approach is very fragile.

(2): Not only does it fail with Hadoop 2, but with every Hadoop version that includes Protobuf 2.5.0 on the classpath.  For `hadoop` CLI this can be made to work by prepending Protobuf 3, or the Ozone FS jar if it includes Protobuf 3 without shading.  However, Hadoop 3.4.1 still includes Protobuf 2.5.0 for YARN and thus MR test [fails](https://github.com/adoroszlai/ozone/actions/runs/17682591631/job/50260853574#step:13:120).

(1): I agree, switching to `ProtobufRpcEngine2` is a huge change.  Using Hadoop's shaded Protobuf may also result in runtime problems if not further shaded.

Other alternatives:

(4): Create an `ozone-thirdparty` repository similar to `ratis-thirdparty`, to bundle all client-side dependencies in a shaded jar.  This has the benefit of reducing build time for Ozone proper, `ozone-thirdparty` needs to be built only when bumping versions or adding/removing dependencies.  However, it needs extra release process.  Since Ozone has more dependencies than Ratis, it may not scale well.  It also requires even more change than (1), but it breaks tight coupling.

(5): Copy `ProtobufRpcEngine` and related classes to Ozone codebase.  (This approach was taken for few other Hadoop components, e.g. `HttpServer`.)  The amount of related items that need to be copied can be reduced by removing unnecessary complexity.  I would even consider copying from Hadoop 3.2, from the time before `ProtobufRpcEngine2` was introduced.  This eliminates complexity of shading Hadoop, but we need to keep an eye on Hadoop and port any important fixes.

**Reply by adoroszlai** [reply]
@jojochuang first step of approach (5) is in #9107, please take a look

**Comment by szetszwo** [comment]
- https://github.com/apache/ozone/pull/9107
- https://github.com/apache/ozone/pull/9112
- https://github.com/apache/ozone/pull/9113
- https://github.com/apache/ozone/pull/9125

I looked at the PRs -- they are all very big involving many backports and new changes.  The forked code probably is high-maintenance.

I have the following suggestion:

(6) move out the ipc code from hadoop-common to a new module, say hadoop-ipc-proto2, and add another new module hadoop-ipc-proto3 which simply depends on hadoop-ipc-proto2 overriding `protobuf.version` to 3.25.8.  Then, Ozone could shade hadoop-ipc-proto3 while hadoop and other projects could remain depending on hadoop-ipc-proto2.

Would it work?

**Reply by adoroszlai** [reply]
Thanks @szetszwo for checking the PRs and the idea about approach (6).

> (6) move out the ipc code from hadoop-common to a new module

`hadoop-common` would need to be split into several modules:
1. base non-RPC code that `hadoop-ipc` relies on
   - configuration
   - metrics
   - I/O
   - tracing
   - utilities
2. `hadoop-ipc`
3. remaining part of `hadoop-common`, including code that uses `hadoop-ipc`

Splitting `hadoop-common` would be a big change, likely only done in Hadoop `trunk`.  Thus we'd need to wait for Hadoop 3.5 release before Ozone can start using it.  We'd also need to drop support for pre-3.5 Hadoop versions.  Given that currently Ozone supports Hadoop 2.10, I'm not sure such a move would be welcome.

Also, note that Hadoop has already upgraded to Protobuf 3, and `ProtobufRpcEngine` has been deprecated.

> I looked at the PRs -- they are all very big involving many backports and new changes.

The PRs are based on each other in series, that's why all look like big changes.  I can update the dependent PRs on squashed commits to clarify which changes belong to which PR.

**Reply by adoroszlai** [reply]
Changes rebased and squashed for clarity:

- 7d5171a40f8fdc8c0892b16e60683c506d3e26e3
- 94beaf4823e7f04bb50d4f64f3e4560e5c4671c1
- 74eae52371651b3c7c3680ffe9c687dac55c891b
- d11f5859c5569d47aa79992edb9b7db3482e3d75

**Reply by szetszwo** [reply]
@adoroszlai , Thanks for the update!  Will review the new commits soon.
